### DISCUSSION TITLE
In place mutation of records where original is no longer used

### DISCUSSION THREAD

**Discussion by qrzychu** [dotnet/csharplang#6796] [Language Ideas]
I've been following the development of Roc lang (https://www.roc-lang.org/) and they have some very clever ideas.

One of them is optimistic in place mutation of records. In cases like:

```charp
record MyRecord(int Id, int Name);

voud someFunc()
{
    var a = new MyRecord(1, "Old Name");
    var b = a with { Name = "New name"};  // a is never used again
}
```

they don't create a copy, instead they modify `a`. Logical end result is the same, but performance is MUCH better.

I am not sure if CLR and Roslyn are able to detect stuff like this, but if yes, maybe it's worth investigating.

Things like:

```chsarp
var result = dbContext.MyRecords.AsNoTracking().Select(r => r with { Id = 0} ).ToArray();
```

could be much faster. Maybe the optimization cuold start with only trivial cases, like local variables in functions, that are allocated in the body no reference is leaving the function body.

**Comment by CyrusNajmabadi** [comment]
In practice this is virtually impossible to optimize for safely. Say that the compiler did optimize this case to mutate in place. However, then the author of the MyRecord type/assembly releases a new version where their record passes itself to some component in it's constructor. Now your assembly is invalid as it will mutate in-place a record that someone else is pointing at.

**Reply by qrzychu** [reply]
yeah, but that could be a JIT optimization or even runtime could detect that if CLR has a specific instruction for record cloning (I doubt it does).

I'm not sure if I understand what exactly are you talking about, because assemblies don't magically get replaced during runtime, and if you restart the program with new version of assembly, JIt kicks in once again (I think).

I am not talking optimizing this when going from C# to IL -> that's not possible I think. More for JIT or AOT (which is JIT just before you run the program :))

**Reply by Richiban** [reply]
In short, this sort of thing is possible in a new language because you can introduce the concepts of ownership and borrowing from the beginning; constructs that are almost impossible to retrofit onto any language, let alone one that's more than 20 years old.

For example, see how basic syntactic methods for tracking ownership are easily defeated:

```cs
record MyRecord(String data);

class C
{
	private readonly List<MyRecord> records = ...;

	public void M()
	{
		var r = new MyRecord("version 1");

		records.Add(r);

		var r2 = r with { data = "version 2" }; // Oops, you've modified the instance of `r` in the list.
	}
}
```

**Reply by qrzychu** [reply]
yeah, but GC for sure has a way to count references. In this case this optimization should not be applied and it's obvious, because you used the `r` variable twice. 

I think even just having "was `r` referenced before the with" check to eliminate cases like this is enough. Add "was r referenced after the with" and "does r leave the scope it was created" should cover quite a lot of ground. You could be as pesimistic as possible with applying this, and still a lot of code would get faster.

I still don't even know how much of difference it would make, I guess I could write some simple benchmarks to compare record copying using `with` and just having a class with a setter.

I will try and do that

**Reply by Richiban** [reply]
> I think even just having "was r referenced before the with" check to eliminate cases like this is enough

But how often is that going to happen? You're introducing a compiler optimisation for occasions where someone has created a record and immediately discarded its value, something that should be caught by any kind of code quality tool or manual code review.

**Reply by CyrusNajmabadi** [reply]
> I am not talking optimizing this when going from C# to IL -> that's not possible I think. More for JIT or AOT (which is JIT just before you run the program :))

This would be a request for the runtime then, not for c#.  Thanks!

**Comment by qrzychu** [comment]
https://github.com/qrzychu/RecordInPlaceMutationTests

So I created some simple benchmarks, and the gain is there: 2-3x faster when mutating:
|                             Method | ItemCount |              Mean |             Error |            StdDev |      Gen0 |      Gen1 |     Gen2 |  Allocated |
|----------------------------------- |---------- |------------------:|------------------:|------------------:|----------:|----------:|---------:|-----------:|
|                CopyMultipleRecords |      1000 |     34,109.856 ns |       484.5953 ns |       453.2907 ns |   49.6216 |         - |        - |   103936 B |
|                      ModifyClasses |      1000 |     27,107.895 ns |       352.9886 ns |       330.1857 ns |   34.1797 |         - |        - |    71936 B |
| MultipleModificationsInARow_Record |      1000 |         15.024 ns |         0.3296 ns |         0.3664 ns |    0.0459 |         - |        - |       96 B |
|  MultipleModificationsInARow_Class |      1000 |          5.666 ns |         0.1113 ns |         0.1041 ns |    0.0153 |         - |        - |       32 B |
|                CopyMultipleRecords |     10000 |    573,340.067 ns |     7,791.3767 ns |     6,906.8525 ns |  211.9141 |  143.5547 |        - |  1039936 B |
|                      ModifyClasses |     10000 |    425,164.225 ns |     8,271.3656 ns |    19,817.6636 ns |  158.2031 |   90.8203 |        - |   719936 B |
| MultipleModificationsInARow_Record |     10000 |         18.464 ns |         0.3847 ns |         0.4430 ns |    0.0459 |         - |        - |       96 B |
|  MultipleModificationsInARow_Class |     10000 |          6.512 ns |         0.1558 ns |         0.1913 ns |    0.0153 |         - |        - |       32 B |
|                CopyMultipleRecords |    100000 | 22,007,577.557 ns | 1,118,663.1573 ns | 3,280,845.3923 ns | 1531.2500 | 1031.2500 | 500.0000 | 10400119 B |
|                      ModifyClasses |    100000 | 11,293,347.922 ns |   680,531.0006 ns | 2,006,561.0095 ns | 1125.0000 |  750.0000 | 328.1250 |  7200058 B |
| MultipleModificationsInARow_Record |    100000 |         16.383 ns |         0.4765 ns |         1.3284 ns |    0.0459 |         - |        - |       96 B |
|  MultipleModificationsInARow_Class |    100000 |          6.737 ns |         0.2017 ns |         0.5852 ns |    0.0153 |         - |        - |       32 B |

**Reply by HaloFour** [reply]
I don't think that's a surprise.  The problem is that it's not an optimization that the compiler can safely make because it has no idea what the constructor of the class or the clone method actually do.  If either copies a reference to the current instance the mutation may be observable.

If you're suggesting that this is something that the runtime can figure out, you can request it on their repo: https://github.com/dotnet/runtime/

But like with so much reference escape analysis the potential performance benefits are likely heavily overshadowed by the overhead of the analysis, and it's just easier to write the code to be more efficient than to try to make the runtime do it.
